posts/Jan/2022-04-29-mapping-boolean-types-in-hibernate-orm-6.adoc

```diff
@@ -0,0 +1,163 @@
+= Mapping Boolean attributes in ORM 6
+Jan Schatteman
+:awestruct-tags: [ "Hibernate ORM" ]
+:awestruct-layout: blog-post
+:icons: font
+:url-hibernate-base: https://docs.jboss.org/hibernate/orm/6.0
+:url-hibernate-javadoc: {url-hibernate-base}/javadocs
+:url-hibernate-ug: {url-hibernate-base}/userguide/html_single/Hibernate_User_Guide.html
+:url-sdk11-javadoc: https://docs.oracle.com/en/java/javase/11/docs/api
+
+
+By default, Hibernate ORM maps a Boolean value to a BOOLEAN column. However, not all databases have a built-in BOOLEAN datatype and not all applications will want to use BOOLEAN (legacy schema e.g.) .
+
+Hibernate provides multiple ways to map Boolean value to other SQL types; this post explores the most common ways.
+
+.Implicit Boolean mapping
+====
+[source, JAVA, indent=0]
+----
+
+    @Basic
+    private boolean aBoolean;
+
+----
+====
+How Hibernate chooses to map a boolean implicitly is determined by a series of cascading checks:
+
+1. It first checks whether the link:{url-hibernate-javadoc}/org/hibernate/cfg/AvailableSettings.html#PREFERRED_BOOLEAN_JDBC_TYPE[hibernate.type.preferred_boolean_jdbc_type] setting has been specified and, if so, uses it;
+2. Next it consults the link:{url-hibernate-javadoc}/org/hibernate/dialect/Dialect.html#getPreferredSqlTypeCodeForBoolean()[Dialect] to see if it reports a preferred SQL type and, if so, uses it;
+3. Otherwise it will use link:{url-sdk11-javadoc}/java.sql/java/sql/Types.html#BOOLEAN[Types.BOOLEAN]
+
+This default, 'implicit' mapping can be modified in several ways.
+
+== JPA compliant, using AttributeConverter
```

Comment:
No need for `JPA compliant, ` here.  It adds no value.

I'd also condense this whole conversation into something like:


```
== AttributeConverter

Using an `AttributeConverter` is a natural way to map a boolean accounting for different storage representations.  
See the link:{url-hibernate-ug}#basic-jpa-convert[User Guide] for details about using `AttributeConverter`.

Hibernate provides 3 built-in `AttributeConverter` implementations for the most common cases:

YesNoConverter:: encodes boolean values as 'Y' or 'N'
TrueFalseConverter:: encodes boolean values as 'T' or 'F'
NumericBooleanConverter:: encodes boolean values as an integer, 1 for true and 0 for false


[[convert]]
.Using AttributeConverter
====
[source, JAVA, indent=0]
----
@Entity
class TheEntity {
    ...

    // this will get mapped to CHAR or NCHAR
    @Basic
    @Convert(converter = org.hibernate.type.YesNoConverter.class)
    boolean convertedYesNo;

    // this will get mapped to CHAR or NCHAR
    @Basic
    @Convert(converter = org.hibernate.type.TrueFalseConverter.class)
    boolean convertedTrueFalse;

    // this will get mapped to TINYINT
    @Basic
    @Convert(converter = org.hibernate.type.NumericBooleanConverter.class)
    boolean convertedNumeric;

    // use our custom converter
    @Basic
    @Convert(converter = CustomBooleanConverter.class)
    boolean convertedCustom;
}

@Converter
public class CustomBooleanConverter implements AttributeConverter<Boolean, Character> {
    ...
}
----
====
```

== Auto-applied AttributeConverter

In the above <<convert,examples>>, the conversions will only be applied to the annotated attribute.

Usually, however, this storage representation will be the same across an entire application or organization.  In such cases, annotating each and every boolean attribute individually is tedious and error prone.

Jakarta Persistence defines a portable way to auto-apply a converter to all boolean types found in the domain model using `jakarta.persistence.Converter#autoApply`.  


[[convert-auto]]
.Auto-applied AttributeConverter
====
[source, JAVA, indent=0]
----
@Converter( autoApply = true )
public class CustomBooleanConverter implements AttributeConverter<Boolean, Character> {
    ...
}
----
====

// We should actually move this discussion about how Hibernate becomes aware of auto-applied 
// converters into the User Guide.  I just noticed we don't really have any content there about that.
// And then:

See the link:{user-guide-url}[User Guide] for details about auto-applied converters.



[[convert-reg]]
== ConverterRegistration

Starting with the imminent 6.1 release, Hibernate also supports registering converters using the new `@ConverterRegistration` annotation.  This annotation is a corollary to the `orm.xml` `<converter/>` element.

Consider Hibernate's built-in `YesNoConverter`, and let's say we want to auto-apply it.  The converter itself is not defined for auto-application (and that would not be appropriate given potential discovery problems).

One option is to explicitly tell Hibernate about it using `orm.xml`:


.<converter/>
====
[source, JAVA, indent=0]
----
<entity-mappings>
    ...
        <converter>org.hibernate.type.YesNoConverter</converter>
    ...
</entity-mappings>
----
====


Another is to use this new annotation, which can be specified at the class or package level.  It's scope is global regardless.

.@ConverterRegistration
====
[source, JAVA, indent=0]
----
@ConverterRegistration( converter=org.hibernate.type.YesNoConverter.class, autoApply=true )
----
====

`ConverterRegistration` will be covered in more detail in the 6.1 version of the User Guide.


// == Hibernate Extensions
// === `@JdbcTypeCode / @JdbcType`


== `@JdbcTypeCode / @JdbcType`

Reply:
Seriously though Steve, it's one thing to correct obvious errors and/or omissions and suggest improvements, and another to rewrite the whole thing. At one point or another you're gonna have to either accept the fact that if someone else wrote something it's not gonna be 100% like you would have done it, or just write it yourself.

Reply:
You are new to writing public blogs.  You asked me for help.  I am suggesting that you write way too much content for a blog post and how you can make the content simpler.  If that irks you, sure.. I'll do it

Reply:
It doesn't irk me, I just feel maybe you're overthinking this a bit. But sure, we can keep chipping away at this until it's reduced to nothing.

Reply:
Ok, I've given my own 'twist' to your suggestions.
For one, I don't see the need for an Auto-applied AttributeConverter section, after all it's about attributeconverters with just an attribute set to true, imo a section of its own suggests that those are some kind of special entities in their own right.
I did not know that (or forgot about) putting a converter in orm.xml auto-applies it, so obviously that has to go in there.